lib/CHeroHandler.h

```diff
@@ -134,6 +134,9 @@ std::vector<std::shared_ptr<Bonus>> SpecialtyBonusToBonuses(const SSpecialtyBonu
 
 class DLL_LINKAGE CHeroClass
 {
+private:
+	// used to store mapping until secondary skills have been loaded
+	std::map<std::string, int> secSkillProbabilityMap;

```

Comment:
Is this needed because hero handler loaded before everything else?
I don't really see if SkillHandler depend on anything else so might be we should just initialize it first?

Reply:
Yes, that's the reason, and you're right - prescribing a load order for handlers would get around the need for this workaround. However, the way I understand it this isn't a good fit with the way things currently work - load order is essentially random (determined by iterator over a map), with anything that requires other things to be loaded first being handled in `afterLoadFinalization`. This lets us deal with cyclic dependencies. There aren't any for SkillHandler, but I think changing the system would require some more discussion.

Reply:
There is special API for such cases `CIdentifierStorage::requestIdentifier` to call functor just after identifier registered
Example from spell handler:
```
	for(const auto & node : json["gainChance"].Struct())
	{
		const int chance = node.second.Integer();

		VLC->modh->identifiers.requestIdentifier(node.second.meta, "faction",node.first, [=](si32 factionID)
		{
			spell->probabilities[factionID] = chance;
		});
	}
```

Reply:
The issue with this approach is that I don't just need the skill id, but also the default chance. And the assumption that once the skill id is available the skill will have loaded is dangerous.

Reply:
This assumption come from `CHandlerBase::loadObject` where registerObject called last and used in almost all handlers

Reply:
What about this (and there might be other similar surprises):
```CHeroHandler::CHeroHandler()
{
	VLC->heroh = this;

	for (int i = 0; i < GameConstants::SKILL_QUANTITY; ++i)
	{
		VLC->modh->identifiers.registerObject("core", "skill", NSecondarySkill::names[i], i);
		VLC->modh->identifiers.registerObject("core", "secondarySkill", NSecondarySkill::names[i], i);
	}
	loadObstacles();
	loadTerrains();
	for (int i = 0; i < GameConstants::TERRAIN_TYPES; ++i)
	{
		VLC->modh->identifiers.registerObject("core", "terrain", GameConstants::TERRAIN_NAMES[i], i);
	}
	loadBallistics();
	loadExperience();
}```

Reply:
You`re right this is yet another legacy